File: Binary search /Single Element in a Sorted Array.cpp
//https://www.naukri.com/code360/problems/unique-element-in-sorted-array_1112654
int singleNonDuplicate(vector<int>& arr)
{
	// Write your code here
	int n= arr.size();

	if(n==1) return arr[0];

	if(arr[0] != arr[1]) return arr[0];
	if(arr[n-1] != arr[n-2]) return arr[n-1];

	int low =1, high=n-2;
	while(low<= high){
		int mid = (low+high)/2;
		if(arr[mid] != arr[mid+1] && arr[mid]!= arr[mid-1]){
			return arr[mid];
		}
		//left side
		if((mid%2==1 && arr[mid]==arr[mid-1]) ||(mid%2==0 && arr[mid]==arr[mid+1])){
			low=mid+1;
		}
		//right side
		else{
			high=mid-1;
		}
	}
	return -1;
}

File: Binary search /Single Element in a Sorted Array.java
//https://leetcode.com/problems/single-element-in-a-sorted-array/description/
class Solution {
    public int singleNonDuplicate(int[] nums) {
    int n= nums.length;

	if(n==1) return nums[0];

	if(nums[0] != nums[1]) return nums[0];
	if(nums[n-1] != nums[n-2]) return nums[n-1];

	int low =1, high=n-2;
	while(low<= high){
		int mid = (low+high)/2;
		if(nums[mid] != nums[mid+1] && nums[mid]!= nums[mid-1]){
			return nums[mid];
		}
		//left side
		if((mid%2==1 && nums[mid]==nums[mid-1]) ||(mid%2==0 && nums[mid]==nums[mid+1])){
			low=mid+1;
		}
		//right side
		else{
			high=mid-1;
		}
	}
	return -1;
    }
}
